### PR TITLE
Add `padding_side` to `pad_sequence` with `"left"` and `"right"` options (`"right"` as default)

### LINKED ISSUE #10536: [feature request] Adding Pre and Post padding functionalities to pad_sequence function
I would like to add **pre** and **post** padding functionalities to ````torch.nn.utils.rnn.pad_sequence```` .
These functionalities haven't been included in the [function](https://pytorch.org/docs/stable/_modules/torch/nn/utils/rnn.html#pad_sequence)

It would be greatly beneficial for people using seq models like LSTM to have sequences created with both **pre** and **post** padding the sequences passed into the args. Other libraries like keras offer these functionalities in their APIs like: ````data = pad_sequences(data, maxlen=50, dtype='float32', padding='pre')````

Besides this, we can add another functionality to allow an argument(with optional value) to know the length of the sequences to be padded to. This is currently done with length of ````sequence[0]````(sequences passed in decreasing order of length).

Thanks!

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by ringohoffman**
Fixes #10536

Reattempt of #61467. Thank you so much to @mskoh52 for your excellent work!

As I was trying to create a more efficient LLM data collator, I realized that `pad_sequence` only supports right padding, even though left padding is a very common format for LLMs, like Llama and Mistral.

The proposed alternative implementation was to use multiple flips, which tends to be 1.5x-2x slower. Instead we can add a [`padding_side` parameter as there is for for Hugging Face tokenizers](https://github.com/huggingface/transformers/blob/9d6c0641c4a3c2c5ecf4d49d7609edd5b745d9bc/src/transformers/tokenization_utils_base.py#L1565), which requires only a very small change in the C++ code.

Here are the benchmarks of the new implementation!

`float32`:

eaaa95ef-9384-45d2-be56-6898bc1d3514

`bool`:

892f32da-8d9a-492b-9507-18d3f0a41e8e

Code:

```python
from __future__ import annotations

import random
import time
from typing import Literal

import numpy as np
import torch


def pad_sequence_with_flips(
    sequences: list[torch.Tensor],
    batch_first: bool = False,
    padding_value: int | float | bool = 0.0,
    padding_side: Literal["left", "right"] | str = "left",
) -> torch.Tensor:
    if padding_side == 'right':
        padded_sequence = torch._C._nn.pad_sequence([t.flatten() for t in sequences], batch_first=batch_first, padding_value=padding_value)
    elif padding_side=='left':
        padded_sequence = torch._C._nn.pad_sequence([t.flatten().flip(0) for t in sequences], batch_first=batch_first, padding_value=padding_value)  # pyright: ignore[reportArgumentType]
        padded_sequence = padded_sequence.flip(int(batch_first))
    else:
        raise ValueError(f"padding_side should be either 'right' or 'left', but got {padding_side}")

    return padded_sequence


sequence_lengths: list[int] = []

flip_left_pad_times: list[float] = []
flip_left_pad_times_std: list[float] = []

left_pad_times: list[float] = []
left_pad_times_std: list[float] = []

RUNS_PER_LOOP: int = 100


for i in range(1, 7):
    sequence_length = i * int(1e6) // 6
    sequence_lengths.append(sequence_length)

    sequences = [torch.randint(0, 2, (random.randint(1, sequence_length),), dtype=torch.bool) for _ in range(64)]

    inner_left_pad_times: list[float] = []
    inner_right_pad_times: list[float] = []

    inner_flip_left_pad_times: list[float] = []
    inner_flip_right_pad_times: list[float] = []

    for _ in range(RUNS_PER_LOOP):

        start = time.perf_counter()
        torch._C._nn.pad_sequence(sequences, batch_first=True, padding_value=False, padding_side="left")
        end = time.perf_counter()
        inner_left_pad_times.append(end - start)

        start = time.perf_counter()
        pad_sequence_with_flips(sequences, batch_first=True, padding_value=False, padding_side="left")
        end = time.perf_counter()
        inner_flip_left_pad_times.append(end - start)

    left_pad_times.append(sum(inner_left_pad_times) / len(inner_left_pad_times))
    left_pad_times_std.append(np.std(inner_left_pad_times))

    flip_left_pad_times.append(sum(inner_flip_left_pad_times) / len(inner_flip_left_pad_times))
    flip_left_pad_times_std.append(np.std(inner_flip_left_pad_times))

    print(f"Sequence Length: {sequence_length}, Left Pad Time: {left_pad_times[-1]}, Left with Flips Pad Time: {flip_left_pad_times[-1]}")


import matplotlib.pyplot as plt

plt.plot(sequence_lengths, left_pad_times, label="new pad_sequence left")
plt.scatter(sequence_lengths, left_pad_times)
plt.errorbar(sequence_lengths, left_pad_times, yerr=left_pad_times_std, linestyle='None', marker='^')

plt.plot(sequence_lengths, flip_left_pad_times, label="old pad_sequence left (2 flips)")
plt.scatter(sequence_lengths, flip_left_pad_times)
plt.errorbar(sequence_lengths, flip_left_pad_times, yerr=flip_left_pad_times_std, linestyle='None', marker='^')

plt.xlabel("Sequence Length")
plt.ylabel("Time (s)")
plt.legend(loc="upper right")

# Sequence Length: 166666, Left Pad Time: 0.06147645162009212, Left with Flips Pad Time: 0.09842291727001794
# Sequence Length: 333333, Left Pad Time: 0.08933195920990329, Left with Flips Pad Time: 0.15597836187991562
# Sequence Length: 500000, Left Pad Time: 0.08863158334006585, Left with Flips Pad Time: 0.15224887342999863
# Sequence Length: 666666, Left Pad Time: 0.10524682551997103, Left with Flips Pad Time: 0.18177212480995877
# Sequence Length: 833333, Left Pad Time: 0.11801802741003485, Left with Flips Pad Time: 0.20821274195001024
# Sequence Length: 1000000, Left Pad Time: 0.131894061660023, Left with Flips Pad Time: 0.23223503091008751
```

cc: @ezyang @jbschlosser @albanD 

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Comment by ringohoffman** [comment]
So it would seem that `test_jit` isn't happy about [`typing.Literal`](https://docs.python.org/3/library/typing.html#typing.Literal), something that was added to Python in 3.8 almost 5 years ago. While I could forego using `Literal`, it does well to document the acceptable values and [is not without precedence in `torch` itself](https://github.com/search?q=repo%3Apytorch%2Fpytorch+%22Literal%5B%22+path%3Atorch%2F**%2F*.py&type=code).

Given JIT's unmaintained status, I wonder what alternatives there might be to removing the `Literal` annotation? Perhaps this JIT regression might be permitted and we could remove this test?

@mikaylagawarecki

**Comment by ezyang** [comment]
I'm happy to take the algorithm change but some bikeshedding on the API may be necessary. The easiest way to bypass the bikeshed is to appeal to some preexisting API that we can copy

**Comment by ringohoffman** [comment]
> The easiest way to bypass the bikeshed is to appeal to some preexisting API that we can copy

I linked the Hugging Face tokenizer API that I based this name off of in the description. Is that what you had in mind?

**Comment by mikaylagawarecki** [comment]
We discussed this at nn triage and were generally ok with the naming of the API being consistent with that of the Huggingface Tokenizer api, but leaving triage review in case there were other thoughts

**Comment by ezyang** [comment]
Triage review looked over the API and we have decided to accept it as is. So we just need a correctness review on this PR.

**Review by ezyang — APPROVED** [review]
